_posts/2020-04-13-Julia-Antipatterns.md

```diff
@@ -0,0 +1,831 @@
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+abstract type AbstractModel end
+
+"""
+    probability_estimate(model, observation::AbstractArray)::Real
+
+For a given `model`, returns the likelihood of the `observation` occurring.
+"""
+function probability_estimate(model::AbstractModel, observation::AbstractArray)
+    error("`probability_estimate` has not been implemented for $(typeof(model))")
+end
+
+## Now a type implementing this API:
+"""
+    GuessingModel <: AbstractModel
+
+A model that just guesses. Not even educated guesses. Just random guessing.
+"""
+struct GuessingModel <: AbstractModel
+end
+
+probability_estimate(guesser::GuessingModel, observation::AbstractMatrix) = rand()
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-cell">
+
+```
+probability_estimate (generic function with 2 methods)
+```
+</div>
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+probability_estimate(GuessingModel(), [1,2,3])
+```
+</div>
+
+**Output:**
+
+    `probability_estimate` has not been implemented for GuessingModel
+
+**What happened?**  
+It sure looks like `GuessingModel` has implemented a `probability_estimate` method.
+The error does not at all help us see what is wrong.
+Astute eyed readers will see what is wrong:
+`GuessingModel()` was incorrectly implemented to only work for `AbstractMatrix`, but it was called with a `Vector`,
+ so it fell back to the generic method for `AbstractModel`.
+But that error message what not informative, and if we hit that deep inside a program we would have no idea what is going on, because it doesn't print the types of all the arguments.
+
+### What to do instead?
+Just don't implement the thing you don't want to implement.
+A `MethodError` indicates this quite well and as shown gives a more informative error message than you will write.
+
+A often missed feature is that you can declare a function without providing any methods.
+This is the ideal way to add documentation for a function that you expect to be overloaded.
+This is done via `function probability_estimate end`.
+As shown. (using `probability_estimate2` to show how it should be done correctly)
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+"""
+    probability_estimate2(model, observation::Vector)::Real
+
+For a given `model`, returns the likelihood of the `observation` occurring.
+"""
+function probability_estimate2 end
+
+probability_estimate2(guesser::GuessingModel, observation::AbstractMatrix) = rand()
+```
+</div>
+
+**Output:**
+
+<div class="jupyter-cell">
+
+```
+probability_estimate2 (generic function with 1 method)
+```
+
+</div>
+
+**Input:**
+
+<div class="jupyter-input jupyter-cell">
+
+```julia
+probability_estimate2(GuessingModel(), [1,2,3])
+```
+</div>
+
+**Output:**
+
+    MethodError: no method matching probability_estimate2(::GuessingModel, ::Array{Int64,1})
+    Closest candidates are:
+      probability_estimate2(::GuessingModel, !Matched::Array{T,2} where T) at In[9]:8
+
```

Comment:
Maybe add a line here to explain the !Match syntax from the compiler, could be useful for beginners